### ISSUE TITLE
Attachment names string treatment is not consistent between different functions

### ISSUE THREAD

**Issue by asibahi** [serenity-rs/serenity#2635]
Serenity version: I am using `poise 0.6.0`

Rust version (`rustc -V`): 1.74

Minimal test case if possible:

```rust
    // inside a command function
    let attachment_name = format!("{}'s_{}.png", "some", "one"); // mind the apostrophe

    let attachment = serenity::CreateAttachment::bytes(image_data, &attachment_name); // 👈 first use here

    let embed = serenity::CreateEmbed::new()
         // other things here
        .attachment(attachment_name); // 👈 second use here

    let reply = poise::CreateReply::default()
        .attachment(attachment)
        .embed(embed);
}
```

the apostrophe in `attachment_name` causes the two functions to give different output somehow. The attachment is no longer *in* the embed.

<img width="598" alt="attachment out of the embed" src="https://github.com/serenity-rs/serenity/assets/17417266/e373d64b-2197-4b46-814a-a8b26d693eb3">

Removing the apostrophe resolved that problem.

<img width="471" alt="attachment inside the embed" src="https://github.com/serenity-rs/serenity/assets/17417266/348624be-ec60-4b16-8908-c07f9988a4ed">

**Comment by ivinjabraham** [comment]
I'll try looking into this

**Comment by ivinjabraham** [comment]
This is caused by a URL sanitation issue. The `url` field in `CreateEmbed(Embed)` shouldn't contain invalid characters. Perhaps `serde` replaces the invalid characters, but somehow that messes up where the attachment should go.

This is a temporary fix:
```rust
    pub fn attachment(self, filename: impl Into<String>) -> Self {
        let mut filename = filename.into();
        filename.retain(|c| c != '\'');
        filename.insert_str(0, "attachment://");
        self.image(filename)
    }
```

Will try and check if there are better ways to ensure `url` is valid and also why it causes the attachment to go out of the embed.

UPDATE: Found that when there are invalid characters, the `Message` struct has attachments, and the embed does not have an image. As expected, when it does not have invalid characters, there are no attachments and the embed as an `EmbedImage`.

**Comment by asibahi** [comment]
Thanks for looking into this. I know of the work around (just not add invalid characters .. duh!). Yes what you're describing is the same behavior I found.

**Comment by jamesbt365** [comment]
Discord converts the names when they include characters it doesn't like, which means that when supplying the attachment name to the embed, it doesn't point to the real attachment anymore.

This is more of a quirk than a bug, and would require filtering the name by the lib, which is currently out of scope incase discord changes in the future.

**Comment by ivinjabraham** [comment]
Maybe we should add a warning in the docs about it? And if there's nothing more to do about this, then perhaps this issue can be closed?

**Comment by asibahi** [comment]
A “fix” would have serenity do Discord’s treatment internally to all strings passed through it.

**Comment by ivinjabraham** [comment]
Is that not what was discussed here:

> would require filtering the name by the lib, which is currently out of scope incase discord changes in the future.

**Comment by jamesbt365** [comment]
Fixed by #2959.
